### PR TITLE
Fix edge case race conditions and clean up lint errors

### PR DESCRIPTION
**Pull request by derekperkins**
inspired by #6

### CODE REVIEW

File: throttler.go [hunk 1]
```diff
@@ -62,26 +64,26 @@ func NewBatchedThrottler(maxWorkers, batchingTotal, batchSize int) *Throttler {
...
 	}
-	t.jobsStarted++
-	t.workerCount++
+	atomic.AddInt32(&t.jobsStarted, 1)
+	atomic.AddInt32(&t.workerCount, 1)
 
-	if t.workerCount == t.maxWorkers {
+	if atomic.LoadInt32(&t.workerCount) == atomic.LoadInt32(&t.maxWorkers) {
+		atomic.AddInt32(&t.jobsCompleted, 1)
+		atomic.AddInt32(&t.workerCount, -1)
```
Comment (josephbergevin) [inline throttler.go:75]: Was this switch of order on purpose?

File: throttler.go [hunk 2]
```diff
@@ -120,18 +126,26 @@ func (te multiError) Error() string {
...
 
+// BatchStartIndex returns the starting index for the next batch. The job count isn't modified
+// until th.Throttle() is called, so if you don't call Throttle before executing this
+// again, it will return the same index as before
 func (t *Throttler) BatchStartIndex() int {
-	return t.jobsStarted * t.batchSize
+	return int(atomic.LoadInt32(&t.jobsStarted) * atomic.LoadInt32(&t.batchSize))
 }
 
+// BatchEndIndex returns the ending index for the next batch. It either returns the full batch size
```
Comment (josephbergevin) [inline throttler.go:136]: Nice comment!

File: throttler_test.go [hunk 3]
```diff
@@ -178,18 +178,19 @@ func TestBatchedThrottler(t *testing.T) {
 		th := NewBatchedThrottler(test.MaxWorkers, len(test.ToBeBatched), test.BatchSize)
 		for i := 0; i < th.TotalJobs(); i++ {
 			go func(tbbSlice []string, expectedSlice []string) {
+				var err error
 				if !reflect.DeepEqual(tbbSlice, expectedSlice) {
-					t.Fatalf("wanted: %#v | got: %#v", expectedSlice, tbbSlice)
+					err = fmt.Errorf("wanted: %#v | got: %#v", expectedSlice, tbbSlice)
```
Comment (josephbergevin) [inline throttler_test.go:183]: Good catch on the test.

**Review by josephbergevin — APPROVED** [review]

File: throttler.go (hunk 1 above)
Comment (derekperkins) [reply throttler.go:75]: Yes, it was causing a race condition here and where it was fixed in #7.

File: throttler.go (hunk 2 above)
Comment (derekperkins) [reply throttler.go:136]: 👍

File: throttler_test.go (hunk 3 above)
Comment (derekperkins) [reply throttler_test.go:183]: Not going to lie, that was `golint` being smart, not me. :)

**Comment by derekperkins** [comment]
Also of note, I switched to Int32 under the hood to use the atomic functions, but cast them back to ints on all the public apis to keep this change BC.

**Pull request merged**